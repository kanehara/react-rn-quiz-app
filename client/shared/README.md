# The shared module for the trivia clients
A module housing common JS code between the web and React Native clients.

## Contains
* Redux actions, reducers, getters
* Apollo client used in redux layer to fetch data
* styled components theme