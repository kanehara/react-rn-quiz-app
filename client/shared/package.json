{
  "name": "shared",
  "version": "0.0.1",
  "scripts": {
    "test": "jest"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "babel-jest": "^24.0.0",
    "jest": "^24.0.0",
    "unfetch": "^4.0.1"
  }
}
