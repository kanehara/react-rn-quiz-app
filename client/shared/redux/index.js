import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { quizReducer } from './quiz'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ 
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) 
    : compose

const persistedReducer = persistReducer({
  key: 'root',
  storage
}, quizReducer)

const store = createStore(
  persistedReducer,
  composeEnhancers(
    applyMiddleware(thunk)
  )
)

const persistor = persistStore(store)

export {
  store,
  persistor
}
