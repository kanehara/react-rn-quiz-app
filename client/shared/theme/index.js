import theme from './theme'
import { ThemeProvider as T } from 'styled-components'
import React from 'react'

const ThemeProvider = ({children}) => <T theme={theme}>{children}</T>

export {
  ThemeProvider
}
