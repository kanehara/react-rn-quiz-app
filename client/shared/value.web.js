export default 'value in web2'
