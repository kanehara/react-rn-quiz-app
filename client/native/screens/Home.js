import React from 'react'
import {Button, ActivityIndicator} from 'react-native'
import Text from 'native/components/Text'
import { H1, H2 } from 'native/components/Header'
import ScreenView from 'native/screens/ScreenView'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'

class Home extends React.Component {
  state = {
    loadingQuiz: false,
    error: null
  }

  componentDidMount () {
    if (this.props.inProgress) {
      this.props.navigation.navigate('quiz')
    } else if (this.props.finished) {
      this.props.navigation.navigate('results')
    }
  }

  handleStartPress = async () => {
    try {
      this.setState({ loadingQuiz: true })
      await this.props.startQuiz()
      this.props.navigation.navigate('quiz')
    } catch (e) {
      this.setState({ error: e, loadingQuiz: false })
    }
  }

  render() {
    return (
      <ScreenView>
        {this.state.loadingQuiz 
          ? <ActivityIndicator />
          : (
            <React.Fragment>
              <H1>Welcome to the Trivia Challenge!</H1>
              <H2>You will be presented with 10 True or False questions.</H2>
              <Text>Can you score 100%?</Text>
              {this.state.loadingQuiz 
                ? <Text>Loading...</Text>
                : <Button onPress={this.handleStartPress} title='BEGIN' />
              }
              {this.state.error && <Text>Something went wrong while fetching questions.</Text>}
            </React.Fragment>
          )
        }
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    inProgress: getters.inProgress(state)
  }),
  dispatch => ({
    startQuiz: () => dispatch(actions.startQuiz())
  })
)(Home)
