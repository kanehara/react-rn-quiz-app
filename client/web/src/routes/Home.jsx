import React from 'react'
import { Ray } from "rayout"
import Button from '../components/Button'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'

export class Home extends React.Component {
  state = {
    loading: false,
    error: null
  }

  componentDidMount () {
    if (this.props.inProgress) {
      this.props.history.replace('/quiz')
    } else if (this.props.finished) {
      this.props.history.replace('/results')
    }
  }

  startQuiz = async () => {
    try {
      this.setState({loading: true})
      await this.props.startQuiz()
      this.props.history.push('/quiz')
    } catch(e) {
      this.setState({ error: e, loading: false })
    }
  }

  render () {
    return (
      <Ray>
        <h1>Welcome to the Trivia Challenge!</h1>
        <h3>You will be presented with 10 True or False questions.</h3>
        <h3>Can you score 100%?</h3>
        {this.state.error 
          ? (
            <React.Fragment>
              <h4>Something went wrong while fetching your quiz questions.</h4>
              <Button onClick={this.startQuiz}>Try again?</Button>
            </React.Fragment>
          )
          : <Button onClick={this.startQuiz}>BEGIN</Button>
        }
      </Ray>
    )
  }
}

const mapStateToProps = state => ({
  finished: getters.finished(state),
  inProgress: getters.inProgress(state)
})

const mapDispatchToProps = (dispatch) => ({
  startQuiz: () => dispatch(actions.startQuiz())
})

export default connect(mapStateToProps, mapDispatchToProps)(Home)
