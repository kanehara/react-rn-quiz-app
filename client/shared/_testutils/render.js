import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

