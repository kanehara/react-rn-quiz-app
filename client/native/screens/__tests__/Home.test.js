import {Home} from '../Home'
import React from 'react'
import { render } from 'shared/_testutils'
import {TouchableHighlight} from 'react-native'

describe('Home', () => {
  const getMockNavigation = () => ({
    navigate: jest.fn()
  })

  it('renders', () => {
    expect(render(<Home />)).toMatchSnapshot()
  })
  
  it('navigates to quiz if in progress', () => {
    const mockNav = getMockNavigation()
    render(<Home navigation={mockNav} inProgress />)
    expect(mockNav.navigate).toHaveBeenCalledWith('quiz')
  })
  
  it('navigates to results if finished', () => {
    const mockNav = getMockNavigation()
    render(<Home navigation={mockNav} finished />)
    expect(mockNav.navigate).toHaveBeenCalledWith('results')
  })
  
  it('navigates to quiz on start', async () => {
    const mockNav = getMockNavigation()
    const startQuizMockPromise = Promise.resolve()
    const startQuizMock = jest.fn(() => startQuizMockPromise)
    const instance = render(<Home navigation={mockNav} startQuiz={startQuizMock} />)
    const button = instance.root.findByType(TouchableHighlight)
    button.props.onPress()

    await startQuizMockPromise
    expect(startQuizMock).toHaveBeenCalled()
    expect(mockNav.navigate).toHaveBeenCalledWith('quiz')
  })
})
