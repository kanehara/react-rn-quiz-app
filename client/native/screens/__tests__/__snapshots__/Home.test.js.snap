// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5faf4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "paddingBottom": 40,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 40,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#292F36",
          "fontSize": 28,
          "fontWeight": "bold",
          "marginBottom": 16,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "textAlign": "center",
        },
      ]
    }
  >
    Welcome to the Trivia Challenge!
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#292F36",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginBottom": 12,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "textAlign": "center",
        },
      ]
    }
  >
    You will be presented with 10 True or False questions.
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#292F36",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginBottom": 12,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "textAlign": "center",
        },
      ]
    }
  >
    Can you score 100%?
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#c5e0de",
          "borderRadius": 10,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 10,
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 10,
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={null}
    >
      <Text
        style={
          Array [
            Object {
              "color": "#292F36",
              "fontSize": 16,
              "fontWeight": "700",
              "marginBottom": 4,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 4,
            },
          ]
        }
      >
        BEGIN
      </Text>
    </View>
  </View>
</View>
`;
